### PR TITLE
Atomic PR

### PR DESCRIPTION
**Pull request by remicalixte**
- basic screen compositing
- primary mode 7 support (lacking transformations)
- shows a black screen when forced blank
- h/v flips for sprites and backgrounds
- fix addressing modes of adc and sbc (:sob:)

### CODE REVIEW

File: core/opcodes.go [hunk 1]
```diff
@@ -416,14 +416,14 @@ func (cpu *CPU) op75() {
 
 func (cpu *CPU) op77() {
 
-	dataLo, dataHi := cpu.admDirectX()
+	dataLo, dataHi := cpu.admBDirectY()
```
Comment (sfluor) [inline core/opcodes.go]: nice catch !

File: core/ppu_bg.go [hunk 2]
```diff
@@ -158,8 +158,13 @@ func (ppu *PPU) bg4vofs(data uint8) {
 
 // tileMapAddress returns the byte address in the VRAM of the tile we are looking for in the tilemap
 // See here: https://wiki.superfamicom.org/backgrounds
-func (bg *bg) tileMapAddress(x uint16, y uint16) uint16 {
-	// TODO: verify that, not sure at all about this
```
Comment (sfluor) [inline core/ppu_bg.go]: So is it the correct formula for getting the tileMapAddress ?

File: core/ppu_bg.go [hunk 3]
```diff
@@ -248,10 +271,14 @@ func (ppu *PPU) colorDepth(background uint8) uint8 {
 		switch background {
 		case 0, 1:
 			return 4
+
+		//TODO offset per tile
+		case 2:
```
Comment (sfluor) [inline core/ppu_bg.go]: quick question: is it possible to render the background 3 in mode 2 ? I thought it was not according to the schema:

```
// 			1   2   3   4
// ======---=---=---=---=
// 0        4   4   4   4
// 1       16  16   4   -
// 2       16  16   -   -
// 3      256  16   -   -
// 4      256   4   -   -
// 5       16   4   -   -
// 6       16   -   -   -
// 7      256   -   -   -
// 7EXTBG 256 128   -   -
```

File: core/ppu_bg.go [hunk 4]
```diff
@@ -273,12 +314,15 @@ func (ppu *PPU) colorDepth(background uint8) uint8 {
...
+			return 2
 		default:
 			panic(fmt.Sprintf(panicString, ppu.backgroundData.screenMode, background+1))
 		}
 	case 7:
 		switch background {
-		case 0:
+		case 0, 1:
```
Comment (sfluor) [inline core/ppu_bg.go]: same question as before for background 2 in mode 7

File: core/ppu_cgram.go [hunk 5]
```diff
@@ -134,16 +134,20 @@ func (ppu *PPU) tileRowColor(tile baseTile, y uint16) [TILE_SIZE]render.BGR555 {
...
+		idx = ppu.colorIndex(tile.addr, uint16(tile.colorDepth), x, y)
+	}
 
 	if idx == 0 {
 		return render.BGR555{Transparent: true}
 	}
 
-	// Sprite colors are stored in the CGRAM starting at palette 8
-	colorWordAddr := 2 * uint16(tile.palette+idx)
+	colorByteAddr := 2 * uint16(tile.palette+idx)
```
Comment (sfluor) [inline core/ppu_cgram.go]: :+1:

File: core/ppu_render.go [hunk 6]
```diff
@@ -20,15 +20,15 @@ func (ppu *PPU) renderLine() {
...
-		ppu.screen.SetPixelLine(ppu.vCounter, ppu.spritesToPixelLine(ppu.oam.intersectingSprites(ppu.vCounter)))
-		//TODO handle background modes and display backgrounds accordingly
-		//We only display BG1 for now
-		ppu.screen.SetPixelLine(ppu.vCounter, ppu.backgroundToPixelLine(0))
+		ppu.paintPixelLine()
 	}
 
 	if ppu.vCounter == ppu.VDisplay()+1 {
-		ppu.renderer.Render(ppu.screen)
+		var screen = ppu.screen
```
Comment (sfluor) [inline core/ppu_render.go]: nit 
```suggestion
		screen := ppu.screen
```

File: core/ppu_render.go [hunk 7]
```diff
@@ -39,23 +39,114 @@ func (ppu *PPU) renderLine() {
...
+// BG2.0    BG2.0    -        -        -        -        -        -
+// OBJ.1    OBJ.1    OBJ.1    OBJ.1    OBJ.1    OBJ.1    OBJ.1    OBJ.1
+// BG3.1    BG3.1b   BG1.0    BG1.0    BG1.0    BG1.0    BG1.0    BG1
+// BG4.1    -        -        -        -        -        -        -
+// OBJ.0    OBJ.0    OBJ.0    OBJ.0    OBJ.0    OBJ.0    OBJ.0    OBJ.0
+// BG3.0    BG3.0a   BG2.0    BG2.0    BG2.0    BG2.0    -        BG2.0p
+// BG4.0    BG3.0b   -        -        -        -        -        -
+// Backdrop Backdrop Backdrop Backdrop Backdrop Backdrop Backdrop Backdrop
+
+func (ppu *PPU) paintPixelLine() {
```
Comment (sfluor) [inline core/ppu_render.go]: :100:

File: core/ppu_bg.go (hunk 2 above)
Comment (remicalixte) [reply core/ppu_bg.go]: It seems to be yes

File: core/ppu_bg.go (hunk 3 above)
Comment (remicalixte) [reply core/ppu_bg.go]: Not directly but it can be used in a weird mode called offset-per-tile

File: core/ppu_bg.go (hunk 4 above)
Comment (remicalixte) [reply core/ppu_bg.go]: Yes it can in mode "7EXTBG" which is mode 7 with an extra background layer

File: core/ppu_bg.go (hunk 3 above)
Comment (sfluor) [reply core/ppu_bg.go]: :+1:

**Pull request merged**